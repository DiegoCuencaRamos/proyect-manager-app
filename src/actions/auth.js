import { auth, GoogleAuthProvider, FacebookAuthProvider, TwitterAuthProvider, GithubAuthProvider } from "../firebase"

// LOGIN
const login = (uid) => ({
    type: 'auth/login',
    payload: uid
})

// Login-middleware
const startGoogleLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(GoogleAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startFacebookLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(FacebookAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startTwitterLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(TwitterAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startGithubLogin = () => {
    return (dispatch, getState) => {
        return auth.signInWithRedirect(GithubAuthProvider)
            .catch(e => console.log(e.message))
    }
}

const startEmailLogin = (email, password, setErrorMessage) => {
    return (dispath, getState) => {
        return auth.signInWithEmailAndPassword(email, password)
            .catch(e => {
                console.log(e.message)
                if(e.message === 'The password is invalid or the user does not have a password.') {
                    setErrorMessage('he password is invalid')
                } else {
                    setErrorMessage(e.message)
                }  
            })
    }
}

const startEmailSignUp = (email, password, setErrorMessage) => {
    return (dispatch, getState) => {
        return auth.createUserWithEmailAndPassword(email, password)
            .catch(e => {
                console.log(e.message)
                setErrorMessage(e.message)
            })
    }
}

// LOGOUT
const logout = () => ({
    type: 'auth/logout'
})

const startLogout = () => {
    return (dipatch, getState) => {
        return auth.signOut()
    }
}

export {
    login,
    startGoogleLogin,
    startFacebookLogin,
    startTwitterLogin,
    startGithubLogin,
    startEmailLogin,
    startEmailSignUp,
    logout,
    startLogout
}