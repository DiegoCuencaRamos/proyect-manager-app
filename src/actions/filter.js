const searchFilterChanged = (searchText = '') => ({
    type: 'filter/serachFilterChanged',
    payload: searchText
})

const sortByFilterChanged = (sortBy) => ({
    type: 'filter/sortByFilterChanged',
    payload: sortBy
})

const visualizeModeFilterChanged = (visualizeMode = 'list') => ({
    type: 'filter/visualizeModeFilterChanged',
    payload: visualizeMode
})

export { searchFilterChanged, sortByFilterChanged, visualizeModeFilterChanged }