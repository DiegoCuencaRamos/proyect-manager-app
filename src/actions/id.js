const setProyectId = id => ({
    type: 'id/setProyectId',
    payload: id
})

const setTaskId = id => ({
    type: 'id/setTaskId',
    payload: id
})

export { setProyectId, setTaskId }