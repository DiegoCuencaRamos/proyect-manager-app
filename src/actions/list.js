const setTextFilter = text => ({
    type: 'list/setTextFilter',
    payload: text
})

const setSortBy = sortBy => ({
    type: 'list/setSortBy',
    payload: sortBy
})

const setCurrentPage = currentPage => ({
    type: 'list/setCurrentPage',
    payload: currentPage
})

const setPageLimit = pageLimit => ({
    type: 'list/setPageLimit',
    payload: pageLimit
})

export {
    setTextFilter,
    setSortBy,
    setCurrentPage,
    setPageLimit
}