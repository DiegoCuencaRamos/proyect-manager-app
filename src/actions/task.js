import database from '../firebase'

// 1. Add task
const addTask = (task) => ({
    type: 'task/addTask',
    payload: task
})

const startAddTask = (taskData = {}) => {
    return (dispatch, getState) => {
        const {
            proyectId = '', 
            name = '', 
            description = '', 
            status = ''
        } = taskData
        const task = { proyectId, name, description, status }

        return database.ref('/tasks').push(task)
            .then(ref => {
                dispatch(addTask({
                    id: ref.key,
                    ...task
                }))
            })     
    }
}

// 2. Edit task
const editTask = (id, updates) => ({
    type: 'task/editTask',
    payload: {
        id,
        updates
    }
})

const startEditTask = (id, updates) => {
    return (dispatch, getState) => {

        return database.ref(`/tasks/${id}`).update(updates)
            .then(() => {
                dispatch(editTask(id, updates))
            })
    }
}

// 3. Remove task
const removeTask = (id) => ({
    type: 'task/removeTask',
    payload: id
})

const startRemoveTask = (id) => {
    return (dispatch, getState) => {

        return database.ref(`/tasks/${id}`).remove()
            .then(() => {
                dispatch(removeTask(id))
            })
    }
}

// 4. Set tasks
const setTasks = (tasks) => ({
    type: 'task/setTasks',
    payload: tasks
})

const startSetTasks = () => {
    return (dispatch, getState) => {

        return database.ref(`/tasks`).once('value')
            .then(dataSnapshot => {
                const tasks = []

                dataSnapshot.forEach(dataChildSnapshot => {
                    tasks.push({
                        id: dataChildSnapshot.key,
                        ...dataChildSnapshot.val()
                    })
                })

                dispatch(setTasks(tasks))
            })
    }
}

export { 
    addTask, 
    startAddTask,
    editTask, 
    startEditTask,
    removeTask,
    startRemoveTask,
    setTasks,
    startSetTasks
}