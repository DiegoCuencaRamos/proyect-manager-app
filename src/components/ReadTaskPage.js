import React from 'react'
import { useSelector } from 'react-redux'
import { useParams } from 'react-router-dom'

const ReadTaskPage = () => {
    const tasks = useSelector(state => state.tasks)
    const { taskId } = useParams()
    const task = tasks.find(task => task.id === taskId)

    return (
        <section>
            <h1>{task.name}</h1>
            <p>{task.status}</p>
            <p>{task.description}</p>
        </section>
)}

export default ReadTaskPage