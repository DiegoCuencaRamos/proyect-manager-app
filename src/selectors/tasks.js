const getVisibleTasks = (proyectId, tasks, { searchText, sortBy }) => {
    return tasks
        // Get filtered tasks by proyect
        .filter(task => task.proyectId === proyectId)
        // Get filtered tasks by searchText
        .filter(task => task.name.toLowerCase().includes(searchText.toLowerCase()))
        // Order tasks by criteria
        .sort((a, b) => {
            if(sortBy === 'name') {
                return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1
            }
        }) 
}

export default getVisibleTasks