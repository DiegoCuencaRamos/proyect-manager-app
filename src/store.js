import { createStore } from 'redux'
import rootReducer from './reducer'

const enhancer = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()

const store = createStore(
    rootReducer,
    /* preloadedState. */
    enhancer,
)

export default store