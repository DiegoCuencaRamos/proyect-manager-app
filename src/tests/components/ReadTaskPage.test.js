import React from 'react'
import { mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router'
import { Provider } from 'react-redux'
import store from '../../store'
import ReadTaskPage from '../../components/ReadTaskPage'
import { addTask } from '../../actions/task'
import { tasks } from '../fixures/tasks'

store.dispatch(addTask(tasks[0]))

test('Should render ReadTaskPage correctly with task data', () => {
    const wrapper = mount(
        <MemoryRouter initialEntries={[{ 
            key: 'static', 
            pathname: '/task/t1'
        }]}>
            <Route path="/task/:taskId">
                <Provider store={store}>
                    <ReadTaskPage />
                </Provider>
            </Route>
        </MemoryRouter>
    )
    expect(wrapper).toMatchSnapshot()
})