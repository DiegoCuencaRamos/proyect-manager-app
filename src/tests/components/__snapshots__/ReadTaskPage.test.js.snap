// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ReadTaskPage correctly with task data 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "static",
        "pathname": "/task/t1",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "static",
            "pathname": "/task/t1",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "static",
          "pathname": "/task/t1",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route
      path="/task/:taskId"
    >
      <Provider
        store={
          Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
      >
        <ReadTaskPage>
          <section
            className="read-page"
          >
            <div
              className="container"
            >
              <h1
                className="read-page__title"
              >
                Task 1
              </h1>
              <p
                className="read-page__status"
              >
                todo
              </p>
              <p
                className="read-page__text"
              >
                This is my task fixure 1
              </p>
            </div>
          </section>
        </ReadTaskPage>
      </Provider>
    </Route>
  </Router>
</MemoryRouter>
`;
